### PR TITLE
Allocate less in limited sparse printing

### PR DESCRIPTION
**Pull request by tkelman**
Avoids unnecessary allocation and reversal of an array of outputs from the method in https://github.com/JuliaLang/julia/pull/22536. Add a few more test cases with dense columns to be sure this is consistent with the output from master. The new tests should pass with or without the rest of the change.

edit: mac travis hung while downloading something, log backed up to https://gist.github.com/6433751028a10ec3d17391cd57bf0c26 and restarted

### CODE REVIEW

File: test/sparse/sparse.jl [hunk 1]
```diff
@@ -1779,6 +1779,21 @@ end
     show(ioc, MIME"text/plain"(), sparse(Int64[1, 1], Int64[1, 2], [1.0, 2.0]))
     @test String(take!(io)) == "1×2 SparseMatrixCSC{Float64,Int64} with 2 stored entries:\n  ⋮"
 
+    # even number of rows
+    ioc = IOContext(IOContext(io, :displaysize => (8, 80)), :limit => true)
```
Comment (Sacha0) [inline test/sparse/sparse.jl]: (Tangent: The docstrings for `IOContext` would benefit from more examples.)

Comment (tkelman) [inline test/sparse/sparse.jl]: very much so, that feature has been underdocumented ever since it was added

Comment (Sacha0) [inline test/sparse/sparse.jl]: Could simplify the nested `IOContext` calls to `ioc = IOContext(io, displaysize = (8, 80), limit = true)` (assuming I understand the docstring correctly and my experiments at the REPL confirm what they appear to)?

File: test/sparse/sparse.jl [hunk 2]
```diff
@@ -1779,6 +1779,21 @@ end
...
+    show(ioc, MIME"text/plain"(), sparse(Int64[1,2,3,4], Int64[1,1,2,2], [1.0,2.0,3.0,4.0]))
```
Comment (Sacha0) [inline test/sparse/sparse.jl]: (Another tangent: That methods of `show` accepting a stream aren't `!`-terminated seems inconsistent with the usual convention?)

File: base/sparse/sparsematrix.jl [hunk 3]
```diff
@@ -171,19 +171,17 @@ function Base.show(io::IOContext, S::SparseMatrixCSC)
...
     if !will_fit
-        count = 0
-        for col = reverse(1:S.n), r = reverse(nzrange(S, col))
-            count += 1
-            push!(lines_bottom, _format_line(r, col))
-            count == print_count && break
+        print(io, "\n  \u22ee")
+        # find the column to start printing in for the last print_count elements
+        nextcol = searchsortedfirst(S.colptr, nnz(S) - print_count + 1)
+        for r = nnz(S) - print_count + 1 : S.colptr[nextcol] - 1
```
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Though not strictly necessary, parens would make the range much easier to parse?

**Review by Sacha0 — APPROVED** [review]
lgtm modulo the minor comments! :) (Caveat: Haven't thought through the potential edge cases.)

**Review by Sacha0 — APPROVED** [review]
lgtm for the additional commit! :)

**Comment by tkelman** [comment]
Will merge tomorrow then if I don't hear otherwise

File: test/sparse/sparse.jl [hunk 4]
```diff
@@ -1807,7 +1807,7 @@ end
     @test String(take!(io)) ==  string("6×3 SparseMatrixCSC{Float64,$Int} with 18 stored entries:\n  [1, 1]",
                                        "  =  1.0\n  [2, 1]  =  1.0\n  ⋮\n  [5, 3]  =  1.0\n  [6, 3]  =  1.0")
 
-    ioc = IOContext(io, :displaysize => (9, 80))
+    ioc = IOContext(io, displaysize = (9, 80))
```
Comment (rfourquet) [inline test/sparse/sparse.jl]: As of today, it seems to be faster and to allocate less to use the `Pair` version rather than keywords; so this commit was more for code-style rather than for reducing allocations right?

Comment (tkelman) [inline test/sparse/sparse.jl]: the iocontext part was stylistic yes, but in test code would the difference be noticeable?

Comment (rfourquet) [inline test/sparse/sparse.jl]: OK thanks; I don't believe it would be noticeable in test code, but in 893a65bddea0bd89c3df2eb0cf8727b1f967738d Jameson did exactly the opposite for other test files, apparently in a performance related PR -- but he may have done it in test files only for stylistic reasons too ;-)
